sw/device/lib/crypto/drivers/entropy_kat.c

```diff
@@ -0,0 +1,169 @@
+  /**
+   * Set to true when the CSRNG instance has been instantiated.
+   */
+  bool instantiated;
+  /**
+   * Set to true when FIPS compliant entropy was provided directly by the
+   * entropy source to instantiate or reseed the CSRNG instance.
+   */
+  bool fips_compliance;
+} entropy_csrng_internal_state_t;
+
+static void entropy_csrng_internal_state_get(
+    entropy_csrng_internal_state_id_t instance_id,
+    entropy_csrng_internal_state_t *state) {
+  // Select the instance id to read the internal state from, request a state
+  // machine halt, and wait for the internal registers to be ready to be read.
+  uint32_t reg = bitfield_field32_write(
+      0, CSRNG_INT_STATE_NUM_INT_STATE_NUM_FIELD, instance_id);
+  abs_mmio_write32(kBaseCsrng + CSRNG_INT_STATE_NUM_REG_OFFSET, reg);
+
+  // Read the internal state.
+  state->reseed_counter =
+      abs_mmio_read32(kBaseCsrng + CSRNG_INT_STATE_VAL_REG_OFFSET);
+
+  for (size_t i = 0; i < ARRAYSIZE(state->v); ++i) {
+    state->v[i] = abs_mmio_read32(kBaseCsrng + CSRNG_INT_STATE_VAL_REG_OFFSET);
+  }
+
+  for (size_t i = 0; i < ARRAYSIZE(state->key); ++i) {
+    state->key[i] =
+        abs_mmio_read32(kBaseCsrng + CSRNG_INT_STATE_VAL_REG_OFFSET);
+  }
+
+  uint32_t flags = abs_mmio_read32(kBaseCsrng + CSRNG_INT_STATE_VAL_REG_OFFSET);
+
+  // The following bit indexes are defined in
+  // https://docs.opentitan.org/hw/ip/csrng/doc/#working-state-values
+  state->instantiated = bitfield_bit32_read(flags, /*bit_index=*/0u);
+  state->fips_compliance = bitfield_bit32_read(flags, /*bit_index=*/1u);
+}
+
+/**
+ * Checks the CSRNG internal state against `expected` values.
+ *
+ * @param csrng A CSRNG handle.
+ * @param expected Expected CSRNG internal state.
+ */
+static status_t check_internal_state(
+    const entropy_csrng_internal_state_t *expected) {
+  entropy_csrng_internal_state_t got = {0};
+  entropy_csrng_internal_state_get(kCsrngInternalStateIdSw, &got);
+  if (memcmp(&got, expected, sizeof(entropy_csrng_internal_state_t)) == 0) {
+    return OK_STATUS();
+  }
+  return INTERNAL();
+}
+
+status_t entropy_csrng_kat(void) {
+  TRY(entropy_csrng_uninstantiate());
+
+  const entropy_seed_material_t kEntropyInput = {
+      .data = {0x73bec010, 0x9262474c, 0x16a30f76, 0x531b51de, 0x2ee494e5,
+               0xdfec9db3, 0xcb7a879d, 0x5600419c, 0xca79b0b0, 0xdda33b5c,
+               0xa468649e, 0xdf5d73fa},
+      .len = 12,
+  };
+  TRY(entropy_csrng_instantiate(
+      /*disable_trng_input=*/kHardenedBoolTrue, &kEntropyInput));
+
+  const entropy_csrng_internal_state_t kExpectedStateInstantiate = {
+      .reseed_counter = 1,
+      .v = {0x06b8f59e, 0x43c0b2c2, 0x21052502, 0x217b5214},
+      .key = {0x941709fd, 0xd8a25860, 0x861aecf3, 0x98a701a1, 0x0eb2c33b,
+              0x74c08fad, 0x632d5227, 0x8c52f901},
+      .instantiated = true,
+      .fips_compliance = false,
+  };
+  if (!status_ok(check_internal_state(&kExpectedStateInstantiate))) {
+    return INTERNAL();
+  }
```

Comment:
Unless you need to translate the status code out of `check_internal_state` from what it naturally returns to something else, this could just be written as:
```
TRY(check_internal_state(...));
```

Reply:
Done.